### ISSUE TITLE
Project coding standards should be documented

### ISSUE THREAD

**Issue by dairiki** [lektor/lektor#888]
Here comes the grumpy guy again.  Please understand that this is not at all meant to be an attack on any particular individual.  But, I think it should be pointed out, in a top-level ticket, rather than in just after-the-fact comments, that repeated adjustment of a stable, multi-developer, open-source project's style requirements creates a lot of friction.  It's questionable whether the benefits, in terms of actually encouraging code maintenance, are commensurate

My thesis:

The project’s coding standards should be documented.  They should not be changed in an impromptu manner.  Changes should be discussed — with the goal being to achieve consensus and buy-in from all contributors.  It generally helps achieve buy-in if the standards are no more restrictive than is really necessary.

## Python is not Javascript

Python, as a language, is somewhat unique in its expressivity and readability.  I'm not totally sure why that is, but because of things like relying on indentation rather than braces, it is hard to write syntactically ugly python code.  Of course, one can write bad code in any language, but in the curly-brace languages, it's easy to make the program control structure hard to read in a way that is nearly impossible to do in Python.  Many of those other languages do benefit from using tools to enforce indentation and other style requirements.  Imposing overly-restrictive style requirements on python code is detrimental to the expressivity that is one of Python's strengths.

For example, black won't let me write:
```
config_ini.write_text(
    "name = Fred\n"
    "[section]\n"
    "foo = bar\n"
)
```
It changes that to the inarguably less readable
```
config_ini.write_text(
    "name = Fred\n" "[section]\n" "foo = bar\n"
)
```

And, does it really improve "readability" to change all the single-quotes to double quotes?  (Sometimes a programmer can put the freedom to choose one or the other to good use.  I tend to use single-quotes for "constant" strings: keys, etc., while using double-quotes for more general strings, e.g. messages.)

While it has a place, I consider the use of tools like black generally harmful.  I think they discourage the careful writing of well-crafted code. See [this blog post](https://luminousmen.com/post/my-unpopular-opinion-about-black-code-formatter) for more on that.

## Unnecessary Friction

In addition to, arguably, at times decreasing the expressivity and readability of Python code, imposing too much pickiness in the style-checkers just creates all kinds of friction for contributors.

Just overnight, apparently, the CI tests (in the master branch) broke due to the release of a new version (2.7) of pylint.  Now everything comes to a halt until someone fixes that.  All these restrictions impose quite a bit of friction.

## Unnecessary Churn

Every few months, it seems another change is made and a bunch of files are reformatted.  See #696 (84 files touched), #823 (79 files), #871 (49 files), #887 (16 files). Reformatting the source code due to style changes on a stable project has many disadvantages.
- Outstanding PRs will no longer apply cleanly (or pass CI tests).  In the past couple weeks, I've had to fix three outstanding PRs for breakage caused by #871 and #887.  This is quite demotivating to contributors.
- It pollutes the commit log, making the git history less useful.  (Look at [this directory](https://github.com/lektor/lektor/tree/master/lektor) and try to determine when the most recent real change to any of the source files was made.  The most recent commit for nearly every file is a "beautification" commit of some sort.)

For the most part, the python code was well-crafted from the beginning.

This is an established, stable code base, currently in use by many, contributed to by dozens over the years.  The python code, for the most part, has always been well-crafted.  (Admittedly, the tests are another issue — they could use some love.)

Arguments such as "consistent style make future debugging easier" are fine, so far as that goes, but, unless one can point out specific issues in the current code base which are truly improved by the changes, is it worth the churn?  If there are one or two places where readability could be significantly improved, just fix those places.

## Community Buy-In

Many of these pylint flags, the use of black, etc. are truly matters of personal preference.  For most of them, there are arguments to be made both ways as to how useful they are in really improving readability or maintainability.  It is rather discouraging to those contributors who happen to feel differently about some of these choices when the PR is made and merged without much discussion.

There is churn in the contributors to this project.  The original author is no longer involved in maintenance.  Even those who were active last year seem to have moved on.  Personally, I feel that it is disrespectful to reformat another author's code without good reason.  (I find it a bit galling that the big blacking PR (#823) appears to have been the last contribution by that particular contributor.)

Getting pickier about style requirements in a public project results in more frustration and more arguments over style requirements.  Is there some bit of the code that is hard to read?  Fix it and move on.  If there really is some pervasive, systematic bit of style that is causing significant problems, please identify it, illustrate that the proposed fix is worth the disruption, and allow for some discussion before pushing it to master.

There are a number of popular and published style guides starting with [PEP 8](https://pep8.org/). 
 There are more specific guides like [google's](https://google.github.io/styleguide/pyguide.html), or [django's](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/) too.  Lots of thought and experience has gone into these.  When starting a new multi-developer project, pick one.  Change it judiciously and deliberately if there is good reason.  For an established, multi-developer project, be even more judicious and deliberate about changing style requirements.

If it is decided that code reformatting is called for, it should be in one big whack, rather than in bits and pieces, so as to minimize the commit log pollution.  And it should be done with lots of warning, to try to avoid horking current PRs, etc.

If one wants to encourage developer participation in the project, it's important that the potential developers feel heard when these things are discussed, and that unnecessary friction to contributing is removed.

## Identify the problem before making the change

It is dismissive to others who care about this code to say, "we should black the code because it makes it more maintainable."  That's fine as far as it goes, and it may be true, but without evidence, it just sounds like corporate-speak.
If there parts of the codebase that would truly be improved by blacking, point them out.  In an established, stable project like this, if there's just one bit that needs fixing, just fix it — there's no need to munge everything.  If the issue is pervasive, point that out.

## My Preference

As you can tell, I believe simpler is better.  My preference, as you may have guessed by now, especially for a mature, stable, open-source project, is PEP 8 (with its "[A Foolish Consistency is the Hobgoblin of Little Minds][cons]".) Along with that, I feel that flake8 is better suited towards open-source, established projects than the pickier pylint. ([Here's more reading](https://www.reddit.com/r/Python/comments/82hgzm/any_advantages_of_flake8_over_pylint/dvai60a/) on that.)

[cons]: <https://pep8.org/#a-foolish-consistency-is-the-hobgoblin-of-little-minds>

Here's a [tweet](https://twitter.com/gvanrossum/status/1227126706089021440?s=20) from some random guy that agrees with me.

**Comment by runfalk** [comment]
Though the PR was created without discussion there was open discussion before we did the "blackening". I think it took a year from the first PR until it was eventually merged. I think all but one active maintainer were supportive and encouraged the change. Black now has a commitment towards not changing the formatting, so once we did the change it is there to stay. I don't think one can make the argument that Black increases churn post merge. Pre black we didn't have a code style (other than what reviewers enforced) though most of it was PEP 8 compliant.

Pinning pylint is something I've voiced before, and if someone makes that PR I'm happy to merge it in. I agree that it's silly that pylint can break CI randomly. I'm also on board with changing pylint to something less strict.

I was one of the supporters of black precisely because it shuts down these kinds of discussions. We picked black as the code style and now we can stop code style discussions in favor of actual features. As a reviewer I never have to think about validating the code style and can focus on the actual logic and readability. It also has the advantage that no one needs to read the code style, since the auto formatter does that for free. To me it doesn't really matter what code style we use, as long as it's consistent and can be auto formatted + validated. Some things black does annoys me, but they happen seldom enough that the workarounds are acceptable to me.

I'm sure you didn't write this because of that particular formatting issue, but you can still get readable code with black by `textwrap` from the standard library:

```python
from textwrap import dedent

ini_content = dedent(
    """
        name = Fred
        [section]
        foo = bar
    """
).lstrip()
config_ini.write_text(ini_content)
```

Your unnecessary churn arguments seems a bit cherry picked to fit your narrative. One of the PRs were closed because it was superseeded by the next one. There even is a file (https://github.com/lektor/lektor/blob/master/.git-blame-ignore-revs) with all commits that were purely stylistic changes so you can ignore them in `git blame`. The first commit is for black and the second is for prettier (JavaScript).

> My thesis:
> 
> The project’s coding standards should be documented. They should not be changed in an impromptu manner. Changes should be discussed — with the goal being to achieve consensus and buy-in from all contributors. It generally helps achieve buy-in if the standards are no more restrictive than is really necessary.

To summarize. Things were discussed in the open. Everyone could voice their opinion for about _a year_. I only recall one vote against black. Changes were made to blacken the code base. Pylint was in place before that, hand has not been openly discussed AFAIK.

I don't think the coding standards need documentation since they are already enforced/autoformatted by black + pylint. They are not changed in an impromptu manner. You have to make a very strong case to change them again.

**Comment by dairiki** [comment]
In my book, the one argument in black's favor is indeed that it should prevent further code style discussion and effort.  But it hasn't really worked in this case — it seems rather to have opened the flood gates to all sorts of pylint and style-related refactorings.  Look at the log messages on the master branch since the blackification.  (I've filtered out my commits, because — well — they're perfect, of course.)

```sh
>>> git log --since=6.months -P --author='^(?!.*dairiki).' --pretty=%s upstream/master \*.py
lint: enable no-self-use
Fix reference cycle in Environment
Merge pull request #879 from dairiki/bug.plugin-distname-message
Race to be the first to use python>=3.6 features
require Python 3.6 and remove Python 3.5 from CI
Fix exception message to make it a bit more helpful.
Reexport types for compatibility (#873)
Minor docstring fixes (#874)
Uniform handling of deferred (descriptor) model data (#789)
more imports at top level
remove useless object inheritance from classes
pylint: only locally disable import-outside-toplevel
remove lektor.uilink
split lektor.environment to reduce cycles
add lektor.constants to reduce import cycles
lektor.types: remove cyclic imports
Fix bug in pluginsystem (#859)
_compat: delete (inline one use of reraise)
_compat: remove itervalues
_compat: remove iteritems
_compat: remove iterkeys
remove type aliases in lektor._compat
tests: move imports to top level
pylint: enable and fix unused-variable
pylint: fix bad-option-value
lint: remove one disable and fix two occurences
cli: remove circular import
Merge pull request #819 from yagebu/drop-python-2
Merge pull request #833 from yagebu/werkzeug-requirement
import io and queue directly and not from _compat
remove 'if PY2'-branches
compat: remove PY2 branch
compat: remove python_2_unicode_compatible
remove deps for Python<3.5
lint: fix code smells pointed out by pylint
setup.py: add python_requires='>3.5'
drop Python 2
test on Python 3.9
deps: allow Werkzeug 0.x and 1
tests: use parametrisation for rsync tests
Merge pull request #830 from arabidopsis/rsync-publish-local
Merge pull request #827 from arabidopsis/limit
tests: fix pylint complaints
allow for local rsync deployment
black test_db.py
offset with _limit None
Beautify code again (#823)
```

Not that all of those are pointless, but for a project that's been in use by many for years, and essentially in maintenance mode, that's quite a bit of pylint and style-adjustment churn.

Personally, were I just starting to consider using Lektor for a project I would look at that (combined with the fact that lektor-website and lektor-atom have sat with tests failing for much of that period) and have serious second thoughts.

In the past few weeks since I've started making a concerted effort to contribute back to the project, there have been four PRs of mine that have been broken by a pylint rules change or similar issue within a day or two of my pushing/updating them.  These incidents were caused by three separate causes (two PRs, one pylint update).  That is very motivation-sapping.

Yes, the standards are documented in various files: the pre-commit config, pylintrc.  But they have not been stable.
```sh
>>> git log --date=short --pretty='%ad %s' upstream/master pylintrc
2021-02-20 lint: enable no-self-use
2021-02-06 remove useless object inheritance from classes
2021-02-06 pylint: only locally disable import-outside-toplevel
2021-02-06 pylint: enable cyclic-import rule
2020-12-06 pylint: enable and fix unused-variable
2020-12-06 pylint: fix bad-option-value
2020-10-25 lint: remove one disable and fix two occurences
2020-10-25 lint: remove some unneeded disables
2020-10-25 pylintrc: one disable per line
2020-10-16 pylintrc: remove disables needed for Python 2
2020-10-09 lint: disable Python3-requiring rules
2020-03-04 fix CI
2019-07-29 [master] Removing bad chars at the beginning of pylintrc.
2018-09-05 [test-fixes] Ignore bad options so pylint in python 2 won't complain about python 3 specific options.
2018-09-05 [test-fixes] Pylint ignore updates
2018-08-12 [pylint-fixes] Adding useless-object-inheritance to ignore list.
2017-06-12 Fix tests (#380)
2016-11-20 Enable pylint (#327)
```

I'm not arguing that black should be dropped (though I wouldn't mind.)  But let's just stop putting so much effort into tweaking the style *requirements* and work on things that actually need fixing.  The code was mostly well-crafted before these changes.

**Comment by dairiki** [comment]
> I'm also on board with changing pylint to something less strict.

I would be very happy to switch from pylint to flake8 if there is consensus to do so.  (If there is, I'd be happy to generate a PR to do so.)

I worry a bit about toe-stepping, however, since this would more-or-less abandon all the recent pytest-related tweaks (mostly listed in my previous comment).  And it would probably mean — if not at first, eventually — cleaning out all the current `# pylint: disable` comments (more churn.)

**Comment by yagebu** [comment]
Yes, the timing of some of these last two PRs was a bit unfortunate with relation to your PRs, sorry about that again.

Except for the `no-self-use` one (where I'd be quite fine with disabling the pylint rule again), these changes were fixing and motivated by actual bugs or issues, so I think it's a bit harsh to brush most of the changes in the last months off as just "style-adjustment".

I'd be in favor of keeping pylint, I find it helpful since it points things that I would try to manually watch out for in a review. It is annoyingly noisy though and doesn't provide many helpful suggestions and the red cross does look quite scary in CI runs, so I understand that that can be off-putting. So I think documentation could be improved to point out that people shouldn't be scared off by pylint warnings. And maybe we can run it in a less CI-breaking way so that its warnings are displayed as just warnings and not errors?

**Comment by xlotlu** [comment]
@dairiki I just dropped by to say: :heart: man. :)

Yes, I'm that one maintainer who opposed the horrible horribleness: https://github.com/lektor/lektor/issues/693#issuecomment-714203961. It's also the reason why I left the project - or rather, the way it's been handled is the reason. Among other things, this one in particular is eyes-rolling-worthy:

> Personally, I feel that it is disrespectful to reformat another author's code without good reason. (I find it a bit galling that the big blacking PR (#823) appears to have been the last contribution by that particular contributor.)

So dear mister, regards for the touch of sanity you brought to the discussion.
